File: monolith/typescript/src/catalog/catalog.service.ts
/**
 * MODULE: catalog — owns `products`.
 *
 * Public API: listProducts, getProduct, reserve, release.
 *
 * No other module may touch the `products` table. If `orders` wants a price it
 * calls getProduct; if it wants stock it calls reserve. Nest enforces the half of
 * this that a framework can: a service is only injectable elsewhere if its module
 * exports it, so this class's public methods are the only surface another module
 * can reach. The other half — that the string 'products' appears in no other file —
 * is discipline, and it is what makes the Session 3 split a mechanical exercise
 * rather than a rewrite.
 */

import { Injectable } from '@nestjs/common';
import { Database } from '../database';
import { InsufficientStock, ProductNotFound } from '../errors';

export interface Product {
  sku: string;
  name: string;
  price_cents: number;
  stock: number;
}

@Injectable()
export class CatalogService {
  constructor(private readonly db: Database) {}

  /**
   * One page of products, plus the cursor for the next.
   *
   * Keyset pagination, not OFFSET. An offset drifts: insert a product while a
   * client is on page 2 and it either sees a row twice or misses one entirely. A
   * cursor is a position in the data, not a count of rows someone else can change.
   *
   * We ask for limit + 1 rows: if the extra one comes back there is another page.
   */
  listProducts(limit = 20, cursor?: string): { items: Product[]; next_cursor: string | null } {
    const rows = cursor
      ? this.db.all<Product>(
          'SELECT sku, name, price_cents, stock FROM products WHERE sku > ? ORDER BY sku LIMIT ?',
          cursor, limit + 1)
      : this.db.all<Product>(
          'SELECT sku, name, price_cents, stock FROM products ORDER BY sku LIMIT ?',
          limit + 1);

    const hasMore = rows.length > limit;
    const items = rows.slice(0, limit);
    return {
      items,
      next_cursor: hasMore && items.length ? items[items.length - 1].sku : null,
    };
  }

  getProduct(sku: string): Product {
    const product = this.db.get<Product>(
      'SELECT sku, name, price_cents, stock FROM products WHERE sku = ?', sku);
    if (!product) throw new ProductNotFound(sku);
    return product;
  }

  /**
   * Takes `qty` units off the shelf and returns the product as it was priced.
   *
   * Nothing here mentions a transaction. Because every service shares one
   * connection, this UPDATE silently joins whatever transaction the caller opened —
   * reserving stock and writing the order become one atomic act, and neither module
   * had to know that.
   *
   * Once catalog is a separate service, that silence becomes a lie. What replaces
   * it is a saga, and a compensating "un-reserve" that has to survive the process
   * crashing between the two calls.
   */
  reserve(sku: string, qty: number): Product {
    const product = this.getProduct(sku);
    if (product.stock < qty) throw new InsufficientStock(sku, qty, product.stock);

    this.db.run(
      'UPDATE products SET stock = stock - ? WHERE sku = ? AND stock >= ?', qty, sku, qty);
    return product;
  }

  /** Puts stock back — used when an order is cancelled. */
  release(sku: string, qty: number): void {
    this.db.run('UPDATE products SET stock = stock + ? WHERE sku = ?', qty, sku);
  }
}

File: monolith/csharp/Catalog/CatalogService.cs
namespace Store.Catalog;

/// <summary>The catalog module's public data.</summary>
public sealed record Product(string Sku, string Name, long PriceCents, long Stock);

/// <summary>One page of products. <c>NextCursor</c> is null on the last page.</summary>
public sealed record ProductPage(IReadOnlyList<Product> Items, string? NextCursor);

/// <summary>
/// MODULE: catalog — owns <c>products</c>.
///
/// Public API: ListProducts, GetProduct, Reserve, Release.
///
/// No other module may touch the <c>products</c> table. If <c>orders</c> wants a price
/// it calls GetProduct; if it wants stock it calls Reserve. C# enforces the half of
/// this that a compiler can: <c>MapProduct</c> below is private, so the only surface
/// another module can reach is the four public methods. The other half — that the
/// string "products" appears in no other file — is discipline, and it is what makes
/// the Session 3 split a mechanical exercise rather than a rewrite.
/// </summary>
public sealed class CatalogService(Db db)
{
    /// <summary>
    /// One page of products, plus the cursor for the next.
    ///
    /// Keyset pagination, not OFFSET. An offset drifts: insert a product while a client
    /// is on page 2 and it either sees a row twice or misses one entirely. A cursor is
    /// a position in the data, not a count of rows someone else can change.
    ///
    /// We ask for limit + 1 rows: if the extra one comes back there is another page.
    /// </summary>
    public ProductPage ListProducts(int limit = 20, string? cursor = null)
    {
        var rows = string.IsNullOrEmpty(cursor)
            ? db.Query("SELECT sku, name, price_cents, stock FROM products ORDER BY sku LIMIT ?",
                MapProduct, limit + 1)
            : db.Query("SELECT sku, name, price_cents, stock FROM products WHERE sku > ?" +
                " ORDER BY sku LIMIT ?", MapProduct, cursor, limit + 1);

        var hasMore = rows.Count > limit;
        var items = hasMore ? rows.Take(limit).ToList() : rows;
        return new ProductPage(items, hasMore && items.Count > 0 ? items[^1].Sku : null);
    }

    public Product GetProduct(string sku) =>
        db.QueryOne("SELECT sku, name, price_cents, stock FROM products WHERE sku = ?",
            MapProduct, sku)
        ?? throw DomainException.ProductNotFound(sku);

    /// <summary>
    /// Takes <paramref name="qty"/> units off the shelf and returns the product as it
    /// was priced.
    ///
    /// Nothing here mentions a transaction. Because every service shares one connection,
    /// this UPDATE silently joins whatever transaction the caller opened — reserving
    /// stock and writing the order become one atomic act, and neither module had to
    /// know that.
    ///
    /// Once catalog is a separate service, that silence becomes a lie. What replaces it
    /// is a saga, and a compensating "un-reserve" that has to survive the process
    /// crashing between the two calls.
    /// </summary>
    public Product Reserve(string sku, long qty)
    {
        var product = GetProduct(sku);
        if (product.Stock < qty)
        {
            throw DomainException.InsufficientStock(sku, qty, product.Stock);
        }

        db.Execute("UPDATE products SET stock = stock - ? WHERE sku = ? AND stock >= ?",
            qty, sku, qty);
        return product;
    }

    /// <summary>Puts stock back — used when an order is cancelled.</summary>
    public void Release(string sku, long qty) =>
        db.Execute("UPDATE products SET stock = stock + ? WHERE sku = ?", qty, sku);

    private static Product MapProduct(Microsoft.Data.Sqlite.SqliteDataReader reader) =>
        new(reader.GetString(0), reader.GetString(1), reader.GetInt64(2), reader.GetInt64(3));
}
